File: tools/repository_revision.mjs
import { execFileSync } from "node:child_process";

function git(args, cwd) {
  return execFileSync("git", args, {
    cwd,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  }).trim();
}

/**
 * Resolve the capture branch tip even when Actions checked out its synthetic
 * pull-request merge commit. Local capture runs have no pull-request event and
 * use HEAD directly.
 */
function captureHead(cwd, mainRef) {
  const candidate = process.env.GITHUB_HEAD_SHA;
  if (candidate) {
    try {
      return git(["rev-parse", "--verify", `${candidate}^{commit}`], cwd);
    } catch {
      // Fall through to the merge-parent detection below.
    }
  }
  if (process.env.GITHUB_EVENT_NAME === "pull_request") {
    try {
      const parents = git(["rev-list", "--parents", "-n", "1", "HEAD"], cwd).split(/\s+/);
      if (parents.length === 3) {
        const [, firstParent, branchParent] = parents;
        if (git(["merge-base", firstParent, mainRef], cwd) === firstParent) return branchParent;
      }
    } catch {
      // Fall back to HEAD so the merge-base command reports the real failure.
    }
  }
  return "HEAD";
}

/** Return the commit shared by the capture branch and the fetched default branch. */
export function resolveRepositoryRevision(cwd, mainRef = "origin/main") {
  return git(["merge-base", captureHead(cwd, mainRef), mainRef], cwd);
}

/** Return the branch tip for optional, explicitly non-authoritative detail. */
export function branchHead(cwd) {
  return git(["rev-parse", "HEAD"], cwd);
}

File: tools/repository_revision.py
"""Resolve capture provenance to a commit that readers can find on main."""

from __future__ import annotations

import argparse
import os
import subprocess
from pathlib import Path


def _git(*args: str, cwd: Path | str | None = None) -> str:
    return subprocess.run(
        ["git", *args],
        cwd=cwd,
        check=True,
        capture_output=True,
        text=True,
    ).stdout.strip()


def _capture_head(cwd: Path | str | None = None, main_ref: str = "origin/main") -> str:
    """Resolve the branch tip behind GitHub's synthetic pull-request merge."""

    candidate = os.environ.get("GITHUB_HEAD_SHA")
    if candidate:
        try:
            return _git("rev-parse", "--verify", f"{candidate}^{{commit}}", cwd=cwd)
        except subprocess.CalledProcessError:
            pass
    if os.environ.get("GITHUB_EVENT_NAME") == "pull_request":
        try:
            parents = _git("rev-list", "--parents", "-n", "1", "HEAD", cwd=cwd).split()
            if len(parents) == 3 and _git("merge-base", parents[1], main_ref, cwd=cwd) == parents[1]:
                return parents[2]
        except subprocess.CalledProcessError:
            pass
    return "HEAD"


def resolve_repository_revision(
    cwd: Path | str | None = None,
    *,
    main_ref: str = "origin/main",
) -> str:
    """Return the grounded commit shared by the capture branch and main.

    Captures may be produced on a branch that is later squash-merged.  The
    branch tip is therefore not durable provenance; its merge-base with the
    fetched default branch is.  Missing ``origin/main`` is an error instead
    of silently recording an unresolvable fallback.
    """

    return _git("merge-base", _capture_head(cwd, main_ref), main_ref, cwd=cwd)


def branch_head(cwd: Path | str | None = None) -> str:
    """Return the branch tip for optional, explicitly non-authoritative detail."""

    return _git("rev-parse", "HEAD", cwd=cwd)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--cwd", type=Path, default=Path.cwd())
    parser.add_argument("--main-ref", default="origin/main")
    args = parser.parse_args()
    print(resolve_repository_revision(args.cwd, main_ref=args.main_ref))
